Controllers/CHumidityController.cpp

```diff
@@ -0,0 +1,195 @@
+                                         etl::string<MAX_STRING_SIZE> name,
+                                         uint32_t run_period_ms,
+                                         SPI_HandleTypeDef *p_spi,
+                                         GPIO_TypeDef *p_slave_select_port,
+                                         uint16_t slave_select_pin)
+    : CController(name, run_period_ms),
+      mp_hw(p_hardware)
+{
+    // TODO Auto-generated constructor stub
+    reset();
+
+    CBME280 humidity_sensor(p_spi, p_slave_select_port, slave_select_pin);
+    m_humidity_sensor = &humidity_sensor;
+}
+
+/**
+ * @brief Run all repeated humidity control activities.
+ *
+ */
+void CHumidityController::run()
+{
+    float power = 0;
+    if (m_power_override == DISABLE_OVERRIDE)
+    {
+        if (m_target_humidity != DISABLE_TARGET)
+        {
+            float actual_humidity = mp_humidity_sensor->getHumidity();
+            if ((MIN_HUMIDITY <= actual_humidity) &&
+                (actual_humidity <= MAX_humidity))
+            {
+                power = m_control_loop.run(m_target_humidity, actual_humidity);
+            }
+        }
+    }
+    else
+    {
+        power = m_power_override;
+    }
+    mp_hw->setHumidifierPower(power);
+}
+
+bool CHumidityController::newCommand(ICommand *p_command,
+                                     IComChannel *p_comchannel)
+{
+    bool b_command_recognised = false;
+    ICommand::command_error_code_t result = ICommand::COMMAND_OK;
+    if (p_command->getName()->compare("?humidity") == 0)
+    {
+        sendStatus(p_comchannel);
+        b_command_recognised = true;
+    }
+
+    if (p_command->getName()->compare("humidity") == 0)
+    {
+        result = setHumidity(p_command);
+        b_command_recognised = true;
+    }
+
+    if (p_command->getName()->compare("heater") == 0)
+    {
+        result = overrideHeater(p_command);
+        b_command_recognised = true;
+    }
+    if (b_command_recognised)
+    {
+        switch (result)
+        {
+            case ICommand::COMMAND_OK:
+                //                p_comchannel->send("OK.\n");
+                break;
+            case ICommand::ERROR_ARG_COUNT:
+                p_comchannel->send("Wrong number of arguments.\n");
+                break;
+            case ICommand::ERROR_OUT_OF_BOUNDS:
+                p_comchannel->send("Argument out of bounds.\n");
+                break;
+            case ICommand::ERROR_TYPE_MISMATCH:
+                p_comchannel->send("Argument type mismatch.\n");
+                break;
+            default:
+                p_comchannel->send("Non-specific error with the command.");
+        }
+    }
+    return b_command_recognised;
+}
+
+void CHumidityController::reset()
+{
+    m_target_humidity = DISABLE_TARGET;
+    m_power_override = DISABLE_OVERRIDE;
+    m_control_loop.reset();
+}
+
+void CHumidityController::sendStatus(IComChannel *p_comchannel)
+{
+    etl::string<MAX_STRING_SIZE> message;
```

Comment:
We are currently discussing how to return responses to be understood by higher layers. This might change drastically in the future.